src/dnaio/_core.pyx

```diff
@@ -298,41 +299,39 @@ cdef bytes create_fastq_record(char * name, char * sequence, char * qualities,
         retval_ptr[cursor] = b"\n"
         return retval
 
-# It would be nice to be able to have the first parameter be an
-# unsigned char[:] (memory view), but this fails with a BufferError
-# when a bytes object is passed in.
-# See <https://stackoverflow.com/questions/28203670/>
 
-ctypedef fused bytes_or_bytearray:
-    bytes
-    bytearray
-
-
-def paired_fastq_heads(bytes_or_bytearray buf1, bytes_or_bytearray buf2, Py_ssize_t end1, Py_ssize_t end2):
+def paired_fastq_heads(buf1, buf2, Py_ssize_t end1, Py_ssize_t end2):
     """
     Skip forward in the two buffers by multiples of four lines.
 
     Return a tuple (length1, length2) such that buf1[:length1] and
     buf2[:length2] contain the same number of lines (where the
     line number is divisible by four).
     """
+    # Acquire buffers. Cython automatically checks for errors here.
+    cdef Py_buffer data1_buffer
+    cdef Py_buffer data2_buffer
+    PyObject_GetBuffer(buf1, &data1_buffer, PyBUF_SIMPLE)
+    PyObject_GetBuffer(buf2, &data2_buffer, PyBUF_SIMPLE)
+
     cdef:
-        Py_ssize_t pos1 = 0, pos2 = 0
         Py_ssize_t linebreaks = 0
-        unsigned char* data1 = buf1
-        unsigned char* data2 = buf2
```

Comment:
With the parameter changed to `const unsigned char[:]`, this needs to be changed to `unsigned char* data1 = &buf1[0]`.

Reply:
Ok, sorry, but by coincidence, there was a discussion on the Cython mailing list about this pattern just today. The problem is that `&buf1[0]` is bounds checked and actually raises an IndexError if `buf1` happens to be an empty string. I think what should work is this:
```
unsigned char* const data1 = buf1
```
If it doesn’t, feel free to revert to the version that explicitly uses the C API.

Reply:
Oh and I think you can also remove the `from cpython.buffer` cimports.

Reply:
Ok, forget it. The weird `const` thing above actually leads to a segfault. Please just revert to [76368ca](https://github.com/marcelm/dnaio/pull/49/commits/76368caab25aaf55cdbfe1765a9dd3c729c885d1).

Reply:
Thanks! I was not happy with this code change for this precise reason. The `cont unsigned char[:]`, just looks weird. Also it does not support `const char[:]` for some reason???

 I am okay with doing Python in Cython, I am okay with doing CPython-API in Cython, but these weird Cython-only constructs always feel off to me. Plus, Cython is not as well maintained as the Python C-API, it is mostly just Stefan Behnel. He does an excellent job, but he can't keep up with the PRs. So I always feel safer not doing Cython-specific calls. It also makes it easier to port code to the C-API or back to pure Python if necessary.

Reply:
The branch was reverted.